### PR TITLE
libc: Move the common implementation of up_testset to libc/machine

### PR DESCRIPTION
**Pull request by xiaoxiang781216**

### CODE REVIEW

File: libs/libc/machine/arch_atomic.c [hunk 1]
```diff
@@ -733,3 +733,26 @@ SYNC_VAL_CMP_SWAP(4, uint32_t)
...
+ * Name: up_testset
+ ****************************************************************************/
+
+#if defined(CONFIG_SPINLOCK) && !defined(CONFIG_ARCH_HAVE_TESTSET)
+spinlock_t up_testset(volatile FAR spinlock_t *lock)
+{
+  irqstate_t flags;
+  spinlock_t ret;
+
+  flags = spin_lock_irqsave(NULL);
```
Comment (pkarashchenko) [inline libs/libc/machine/arch_atomic.c:747]: @xiaoxiang781216 are you sure in this change? The first look makes me thinking that change from `up_irq_save` to `spin_lock_irqsave` will make a recursion.

Comment (pkarashchenko) [reply libs/libc/machine/arch_atomic.c:747]: I mean `up_testset` -> `spin_lock_irqsave` -> `spin_lock` -> `up_testset`

Comment (xiaoxiang781216) [reply libs/libc/machine/arch_atomic.c:747]: This call chain only happen in SMP case, but SMP require arch provide up_testset implement(ARCH_HAVE_TESTSET):
https://github.com/apache/incubator-nuttx/blob/master/sched/Kconfig#L275
so, the implementation in here is only used for single core.

Comment (pkarashchenko) [reply libs/libc/machine/arch_atomic.c:747]: So why we can't keep up_irq_save?

**Comment by pkarashchenko** [comment]
What about this code:
```  /* Loop until we successfully take the spinlock (i.e., until the previous
   * state of the spinlock was SP_UNLOCKED).
   * NOTE that the test/set operation is performed via boardctl() to avoid a
   * variety of issues.  An option might be to move the implementation of
   * up_testset() to libs/libc/machine.
   */

  do
    {
#ifdef CONFIG_BUILD_FLAT
      ret = up_testset(&lock->sp_lock) == SP_LOCKED ? 1 : 0;
#else
      ret = boardctl(BOARDIOC_TESTSET, (uintptr_t)&lock->sp_lock);
#endif
    }
  while (ret == 1);
```
should it be modified now?

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (xiaoxiang781216) [reply libs/libc/machine/arch_atomic.c:747]: just align with the code at:
https://github.com/apache/incubator-nuttx/pull/7649/files#diff-60c271175313a6344e1faeba2708393238fd3caf3cf282b3f247d9a88699cba9R67

**Comment by xiaoxiang781216** [comment]
Can't since up_testset stil can't be called from userspace in protected and kernel mode since up_irqsave require the privilege right.

File: include/nuttx/spinlock.h [hunk 2]
```diff
@@ -105,28 +105,7 @@ typedef uint8_t spinlock_t;
  *
  ****************************************************************************/
 
-#if defined(CONFIG_ARCH_HAVE_TESTSET)
 spinlock_t up_testset(volatile FAR spinlock_t *lock);
```
Comment (pkarashchenko) [inline include/nuttx/spinlock.h:108]: could you please fix a comment near `spin_islocked`? and modify to `/* bool spin_islocked(FAR spinlock_t *lock); */`

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (pkarashchenko) [reply libs/libc/machine/arch_atomic.c:747]: Sorry, that link does not redirect me to anything meaningful

Comment (xiaoxiang781216) [reply libs/libc/machine/arch_atomic.c:747]: Here:
image

File: include/nuttx/spinlock.h (hunk 2 above)
Comment (xiaoxiang781216) [reply include/nuttx/spinlock.h:108]: Done.

**Review by jerpelea — APPROVED** [review]

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (pkarashchenko) [reply libs/libc/machine/arch_atomic.c:747]: Yeah, those are atomic values, but spinlock itself relies at up_testest, so using spinlock inside seems to be odd at least to me.
I fully understand that spinlock can be implemented on top of atomic compare-and-swap, but this is not the case here. Also the macro for a single CPU system will end up into disabling of interrupts as it was in the code before this change.
@masayuki2009 what is your opinion here?

File: libs/libc/machine/arch_atomic.c [hunk 3]
```diff
@@ -733,3 +733,26 @@ SYNC_VAL_CMP_SWAP(4, uint32_t)
 SYNC_VAL_CMP_SWAP(8, uint64_t)
 
 #endif /* __clang__ */
+
+/****************************************************************************
+ * Name: up_testset
+ ****************************************************************************/
+
+#if defined(CONFIG_SPINLOCK) && !defined(CONFIG_ARCH_HAVE_TESTSET)
+spinlock_t up_testset(volatile FAR spinlock_t *lock)
```
Comment (masayuki2009) [inline libs/libc/machine/arch_atomic.c:742]: @xiaoxiang781216 
Can we test the above condition with existing defconfig?

File: libs/libc/machine/arch_atomic.c (hunk 1 above)
Comment (masayuki2009) [reply libs/libc/machine/arch_atomic.c:747]: >@xiaoxiang781216 are you sure in this change? The first look makes me thinking that change from up_irq_save to spin_lock_irqsave will make a recursion.

I think so too.

Comment (xiaoxiang781216) [reply libs/libc/machine/arch_atomic.c:747]: Ok, let's keep it as inline function.

**Pull request closed without merging**